### PR TITLE
Add Inflector component (from StringUtil of PropertyAccess)

### LINKED ISSUE #18166: [PropertyInfo] Singularize array mutator in ReflectionExtractor
`ReflectionExtractor` currently supports array mutators (`add`, `remove`) but it's not very useful as the singular method name is not being checked.

If we can use `doctrine/inflector` to singularize the property name, we'll be able to pick up a lot more useful information in many cases. I'll open a PR if this is desirable (looking at the `composer.json` of Symfony components, I understand that it has to be an optional dependency).

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | no |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | N/A |

As proposed by @dunglas here: https://github.com/symfony/symfony/issues/18166#issuecomment-196707518

This will help us fix #18166

### CODE REVIEW

File: src/Symfony/Component/Inflector/phpunit.xml.dist [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.1/phpunit.xsd"
```
Comment (stof) [inline src/Symfony/Component/Inflector/phpunit.xml.dist]: should be ``current``, not ``4.1``

**Comment by stof** [comment]
We need to keep some legacy test for the StringUtil class

File: src/Symfony/Component/Inflector/LICENSE [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2012-2016 Fabien Potencier
```
Comment (teohhanhui) [inline src/Symfony/Component/Inflector/LICENSE]: The year is 2012 based on 7837f50c735d4be26f8c47853a1a64e309ae0ea9

File: src/Symfony/Component/Inflector/Inflector.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ */
+
+namespace Symfony\Component\Inflector;
+
+/**
+ * Converts words between singular and plural forms.
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ */
+final class Inflector
```
Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: It it's a new component, we should probably introduce an interface, an mark it `@internal` until we add at least

Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: We should probably mark the whole component as `@internal` for now.

File: src/Symfony/Component/Inflector/Inflector.php [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        array('s', 1, true, true, ''),
+
+        // chateaux (chateau)
+        array('xuae', 4, false, true, 'eau'),
+    );
+
+    /**
+     * This class should not be instantiated.
+     */
+    private function __construct()
```
Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: Ins't it the good time to remove the static interface?

**Comment by teohhanhui** [comment]
@stof
> We need to keep some legacy test for the StringUtil class

Should the test be duplicated?

**Comment by stof** [comment]
> Should the test be duplicated?

no need to duplicate the whole tests (as the new component is tested). But we at least need a simple test, which would catch mistakes in the BC class

**Comment by javiereguiluz** [comment]
What about using [Doctrine Inflector](https://github.com/doctrine/inflector) library instead of creating a new component?

**Comment by teohhanhui** [comment]
@javiereguiluz To me that'd be preferable too. But we need to use the same inflector for both PropertyAccess and PropertyInfo, according to @dunglas: https://github.com/symfony/symfony/issues/18166#issuecomment-196707518

And if we switch both to the Doctrine Inflector, how should we handle the BC break?

**Comment by iltar** [comment]
I think it would be preferable for the community to use only 1 inflector library. Would it be an idea to use the doctrine variant instead and start patching that one to cover more use-cases when needed?

**Comment by teohhanhui** [comment]
How do we handle the BC break? Or would it not be considered a BC break? :stuck_out_tongue_winking_eye:

**Comment by iltar** [comment]
@teohhanhui I don't see why it would be. You can leave the deprecation as is and leave the tests as is. If the tests fail, you can patch those missing use-cases in the doctrine inflector. The only BC break would be tests failing because the doctrine inflector misses cases, but you will find out by hooking it up.

**Comment by teohhanhui** [comment]
@iltar I'm thinking BC breaks for applications due to different result of the singularization. It seems impossible for us to catch them all...

**Comment by teohhanhui** [comment]
But alas, it'd be for the best. What do you think, @dunglas?

**Comment by javiereguiluz** [comment]
> I'm thinking BC breaks for applications due to different result of the singularization.

How could be any difference if both libraries supposedly do the same: transform English words into/from singular/plural?

**Comment by stof** [comment]
> How could be any difference if both libraries supposedly do the same: transform English words into/from singular/plural?

Given that the conversion rules are quite complex, they may behave differently for some cases (common rules should work fine, but the component would be useless if the rule was as simple as removing the final ``s``)

**Comment by teohhanhui** [comment]
If the application previously used the wrong singular form which matches what is produced by `StringUtil::singularify`, it'll stop working once we switch to Doctrine Inflector (given that the singular form is now correct). And vice versa. What used to be correct could now be wrong.

It's not really possible for us to ensure consistent before/after results. Now the question is whether this is part of our BC promise (of course, surprises like this would still be unwelcome and annoying for an application developer).

**Comment by iltar** [comment]
Luckily we have tests to verify the symfony variant still works, if they pass, no problem. If they fail: patch the inflector on doctrine's side.

**Comment by stof** [comment]
The compatibility test between both library would be as follow:
- run the Symfony StringUtil testcases against the doctrine/inflector implementation
- run the doctrine/inflector testcases against the StringUtil implementation (for relevant tests of course, as doctrine/inflector contains more methods than just singularify).

Test failures there would show us known incompatibilities (we may still have incompatibilities in things not covered by any of the testsuites). This would help us to make the decision.

**Comment by teohhanhui** [comment]
> run the Symfony StringUtil testcases against the doctrine/inflector implementation

There are many failures: https://travis-ci.org/teohhanhui/symfony/jobs/117700871#L4206-L4791

And there's a bigger problem... Doctrine Inflector's singularize doesn't support multiple singular forms (even though StringUtil mostly returned bogus results, but there are many valid cases for this...)

**Comment by teohhanhui** [comment]
After clean-up: https://travis-ci.org/teohhanhui/symfony/jobs/117721231#L4099-L4348

There are still a number of failures, and a few that we have no real way of fixing in the Doctrine Inflector (multiple words with the same plural form):

- axes => axe, ax, axis
- bases => base, basis

**Comment by stof** [comment]
> There are still a number of failures, and a few that we have no real way of fixing in the Doctrine Inflector (multiple words with the same plural form):

And this is the difference between the doctrine inflector and the Symfony StringUtil: the doctrine inflector aims to provide a single singular string (which make sense as their primary usage of the inflector is generating method names during code generation, so they must settle on a single singular), while the goal of the StringUtil is to find all potential singular, to be able to call the method named by the developer with one of them.

and this difference of goal of the project is probably the main reason why we won't be able to switch to doctrine/inflector (as they won't change the focus of the Doctrine project just to allow us to use it)

**Comment by teohhanhui** [comment]
@stof It looks like I have to continue with this PR then...

File: src/Symfony/Component/Inflector/Inflector.php (hunk 3 above)
Comment (teohhanhui) [inline src/Symfony/Component/Inflector/Inflector.php]: If we mark this as `@internal`, what's the migration path for existing applications using `StringUtil::singularize`?

File: src/Symfony/Component/Inflector/Inflector.php (hunk 4 above)
Comment (teohhanhui) [inline src/Symfony/Component/Inflector/Inflector.php]: That means we'll need to construct the Inflector and inject it where it's needed. What's the expected benefit?

File: src/Symfony/Component/Inflector/phpunit.xml.dist (hunk 1 above)
Comment (teohhanhui) [inline src/Symfony/Component/Inflector/phpunit.xml.dist]: Every other component uses `4.1` though. Perhaps this should be addressed in a separate PR (if it's an issue)?

File: src/Symfony/Component/Inflector/Inflector.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: Not necessary to inject it. You can construct it in a constructor for instance. The benefit is extensibility. Thanks to the interface you can replace the built in inflector with a custom one.

File: src/Symfony/Component/Inflector/Inflector.php (hunk 3 above)
Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: As StringUtil is part of Symfony, it can use the internal class. There is no problem here.

Comment (teohhanhui) [inline src/Symfony/Component/Inflector/Inflector.php]: We are deprecating `StringUtil::singularify`, aren't we? Since it was a public method, what should users switch to when they upgrade?

Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: You're right. To avoid the problem with an incomplete interface, maybe can we introduce a `PluralizerInterface` (and later we'll add a `SingularizerInterface`).

That way, there is no need to mark the component `@internal`.

File: src/Symfony/Component/Inflector/Inflector.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: Another immediate benefit: the ability to use custom inflectors supporting non-english languages.

File: src/Symfony/Component/Inflector/Inflector.php (hunk 3 above)
Comment (stof) [inline src/Symfony/Component/Inflector/Inflector.php]: I would not add an interface yet, until we figure out what are the real needs for these interfaces

Comment (dunglas) [inline src/Symfony/Component/Inflector/Inflector.php]: @stof what do you think of this use case: https://github.com/symfony/symfony/pull/18260#discussion_r57116826 (just asking, I don't need that for now)

File: src/Symfony/Component/Inflector/Inflector.php (hunk 4 above)
Comment (teohhanhui) [inline src/Symfony/Component/Inflector/Inflector.php]: We have to decide on the scope of this component.

**Comment by fabpot** [comment]
@teohhanhui Do you think you can finish (not sure what's left) this PR is the next couple of days? I'd like to get this in for 3.1 as it seems quite easy to acheive. Marking the whole component as being internal also gives us the opportunity to add more or even drop it if we find a better replacement in the meantime.

**Comment by teohhanhui** [comment]
What else do I need to do, other than adding `@internal` to the classes and methods?

**Comment by teohhanhui** [comment]
Marked as internal and added a disclaimer in the README

**Comment by teohhanhui** [comment]
Rebased.

File: src/Symfony/Component/Inflector/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+Inflector Component
+======================
```
Comment (fabpot) [inline src/Symfony/Component/Inflector/README.md]: Extra `=`s here

File: src/Symfony/Component/PropertyAccess/StringUtil.php [hunk 6]
```diff
@@ -143,80 +39,13 @@ private function __construct()
...
-                    break;
-                }
-            }
-        }
-
-        // Convert teeth to tooth, feet to foot
-        if (false !== ($pos = strpos($plural, 'ee')) && strlen($plural) > 3 && 'feedback' !== $plural) {
-            return substr_replace($plural, 'oo', $pos, 2);
-        }
+        @trigger_error('StringUtil::singularify() is deprecated since version 3.1 and will be removed in 4.0.', E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/StringUtil.php]: Can you add at the end another sentence to tell the user to use the new class? `Use .... instead.`

**Comment by fabpot** [comment]
:+1: (after the minor comments have been fixed)

File: src/Symfony/Component/PropertyAccess/StringUtil.php (hunk 6 above)
Comment (teohhanhui) [inline src/Symfony/Component/PropertyAccess/StringUtil.php]: We're marking the new class as internal. Should the user still use it?

Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/StringUtil.php]: That's a good point. I think that nobody should be using this one as well, so we should give an alternative. Saying that something is deprecated without a replacement is not good.

Comment (teohhanhui) [inline src/Symfony/Component/PropertyAccess/StringUtil.php]: Perhaps `doctrine/inflector`.

File: src/Symfony/Component/PropertyAccess/StringUtil.php [hunk 7]
```diff
@@ -11,122 +11,18 @@
...
 namespace Symfony\Component\PropertyAccess;
 
+use Symfony\Component\Inflector\Inflector;
+
 /**
  * Creates singulars from plurals.
  *
  * @author Bernhard Schussek <bschussek@gmail.com>
+ *
+ * @deprecated Deprecated since version 3.1, to be removed in 4.0. Use {@see \Doctrine\Common\Inflector\Inflector} instead.
```
Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/StringUtil.php]: Telling people to use Doctrine Inflector does not make sense here as the feature is not the same. And in fact, we are creating a new component exactly for this reason: we cannot use Doctrine inflector. So, the recommendation should be to use the new component instead.

**Comment by dunglas** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @teohhanhui.

**Comment by flip111** [comment]
Is/can this component be used for https://github.com/symfony/symfony/issues/18149 ?
